File: packages/bot-engine/src/strategies/EMAScalp.js
import { EMA } from 'technicalindicators';

/**
 * EMA Scalp Strategy — EMA 3/8 Cross
 * เข้าเร็วมาก ตอบสนองทุก candle เหมาะกับ 1m-5m
 */
export const EMAScalpStrategy = {
  name: 'EMA_SCALP',
  description: '⚡ EMA 3/8 Scalping — เข้าเร็ว ออกเร็ว',

  compute: (closes, params = {}) => {
    const fast = params.fastPeriod || 3;
    const slow = params.slowPeriod || 8;

    const emaFast = EMA.calculate({ period: fast, values: closes });
    const emaSlow = EMA.calculate({ period: slow, values: closes });

    if (emaFast.length < 2 || emaSlow.length < 2) return 'NONE';

    const [pF, cF] = [emaFast.at(-2), emaFast.at(-1)];
    const [pS, cS] = [emaSlow.at(-2), emaSlow.at(-1)];

    if (pF <= pS && cF > cS) return 'LONG';
    if (pF >= pS && cF < cS) return 'SHORT';

    return 'NONE';
  },

  describe: (signal, params = {}) => {
    const fast = params.fastPeriod || 3;
    const slow = params.slowPeriod || 8;
    if (signal === 'LONG') return `⚡ EMA ${fast} ตัดขึ้นเหนือ EMA ${slow} — Scalp LONG`;
    if (signal === 'SHORT') return `⚡ EMA ${fast} ตัดลงใต้ EMA ${slow} — Scalp SHORT`;
    return '';
  },

  getDiagnostic: (closes, params = {}) => {
    const fast = params.fastPeriod || 3;
    const slow = params.slowPeriod || 8;
    if (closes.length < slow) return `กำลังรวบรวมข้อมูล EMA (${fast}/${slow})...`;

    const emaFast = EMA.calculate({ period: fast, values: closes });
    const emaSlow = EMA.calculate({ period: slow, values: closes });

    const cF = emaFast.at(-1);
    const cS = emaSlow.at(-1);
    const gap = (cF - cS).toFixed(4);
    const trend = cF > cS ? 'ขาขึ้น' : 'ขาลง';

    return `EMA ${fast}: ${cF?.toFixed(4)} | EMA ${slow}: ${cS?.toFixed(4)} | ช่องว่าง: ${gap} (${trend}) — รอ Cross`;
  }
};

File: packages/strategy-ai/strategies/ema_scalp.py
import pandas as pd
import ta
from base_strategy import BaseStrategy


class EMAScalpStrategy(BaseStrategy):
    """EMA 3/8 Scalping — powered by ta"""

    def populate_indicators(self, df: pd.DataFrame, params: dict) -> pd.DataFrame:
        p = params or {}
        fast_p = int(p.get("fastPeriod", 3))
        slow_p = int(p.get("slowPeriod", 8))
        df["ema_fast"] = ta.trend.ema_indicator(df["close"], window=fast_p)
        df["ema_slow"] = ta.trend.ema_indicator(df["close"], window=slow_p)
        return df

    def populate_signals(self, df: pd.DataFrame, params: dict) -> pd.DataFrame:
        prev_fast = df["ema_fast"].shift(1)
        prev_slow = df["ema_slow"].shift(1)

        df["signal"] = "NONE"
        df.loc[(prev_fast <= prev_slow) & (df["ema_fast"] > df["ema_slow"]), "signal"] = "LONG"
        df.loc[(prev_fast >= prev_slow) & (df["ema_fast"] < df["ema_slow"]), "signal"] = "SHORT"
        return df

    def get_metadata(self) -> dict:
        return {"name": "EMA_SCALP", "description": "EMA 3/8 Scalp", "version": "3.0.0"}
